### ISSUE TITLE
Problems while not using setTimeout

### ISSUE THREAD

**Issue by denysvitali** [abbr/deasync#11]
Hi everyone,
I've been trying this awesome module for quite a long time, but just today figured out that it has some issues if no timer is set (for example the callback comes from a custom function that makes a call to the request() module)

```
customRequest("GET", "/",
        {}, function(err, res, body)
        {
            var json = JSON.parse(body);
            setTimeout(function(){done = true;}, 1); //workaround
        });
while (!done)
        {
            deasync.runLoopOnce();
        }
```

I've just figured out that when I don't set a timeout but i set the "done" variable to true the deasync module keeps waiting. Seems like deasync.runLoopOnce() stucks since if I debug the loop by adding a `console.log("loop()");` after the done is set to true it keeps waiting, till the done is set by a setTimeout function.

I'm currently using node v.0.12.0
```
denys@denys-pc:~/$ node -v
v0.12.0
```

**Comment by abbr** [comment]
Your code is not self-contained for me to reproduce the problem (i.e. customRequest and done are not defined).
Following code, similar to your except replacing customRequest with setTimeout works for me, so I can argue that the problem exists in customRequest, not in deasync:

```
var done = false;
var data;
setTimeout(function cb(){
  data = 'hello';
  done = true;
},100);
while(!done) {
  require('deasync').runLoopOnce();
}
// output hello as expected
console.log(data);
```

**Comment by denysvitali** [comment]
The problem actually is that if I have a function (let's assume the name is graphRequest for this example) that actually doesn't use a setTimeout and I try something like

This doens't work, but `"Request done",true` is displayed and "Done" is never displayed
```
var done = false;
graphRequest("GET", this.id,
        {}, function(err, res, body)
        {
            var data = JSON.parse(body);
            done = true;
           console.log("Request done",done);
        })
        while (!done)
        {
            deasync.runLoopOnce();
        }
       console.log("Done");
```

This actually do work, because the callback contains a timeout.
Don't know why, just found this workaround.
The thing I couldn't explain to myself is why actually even if done is true, the while stops but deasync.runLoopOnce() get stuck and don't let "Done" get displayed.

```
var done = false;
graphRequest("GET", this.id,
        {}, function(err, res, body)
        {
            var data = JSON.parse(body);
            setTimeOut(function(){ done = true; console.log("Request done",done); },1);
        })
        while (!done)
        {
            deasync.runLoopOnce();
        }
       console.log("Done");
```

**Comment by denysvitali** [comment]
Nevermind, just tested out in clean code and it actually works.
Strange behavior. I'll try to find out the reason.

I close this for now...

**Comment by denysvitali** [comment]
```
var deasync = require('deasync');
var request = require('request');
var express = require('express');
var app = express();


/* MODELS */

function Foo(id)
{
    this.id = id;
    this.bar = function()
    {
        var done = false;
        var data;

        request("http://google.com/", function(err, res, body)
        {
            console.log("done1_in");
            //setTimeout(function(){done = true;}, 1); //workaround
            done = true;
        });

        while (!done)
        {
            deasync.runLoopOnce();
        }

        console.log("done1_out");
        return;
    };
}

function init()
{
    app.listen(8080);
    testfunc();
}


function testfunc()
{
    var data = null;
    var done = false;

    request("http://google.com/", function(err, res, body)
    {
        var foo = new Foo(1);
        foo.bar();
        done = true;
        console.log("done2_in");
    });
    while (!done)
    {
        deasync.runLoopOnce();
    }
    console.log("done2_out");
}

init();
```

Found out the problem:
actually if app.listen function of express module is over the deasync chain casuses problems.

Could you check please? 

The result is:
```
done1_in
```

Instead of: 
```
done1_in
done1_out
done2_in
done2_out
```

**Comment by abbr** [comment]
I can reproduce the problem using your code. But if I replace the async `request` call in `this.bar` function with `setTimeout` then everything works as expected. It appears that `request` lost its context, i.e. the `done` it sets is not the one defined in `this.bar` which deasync relies upon. So again I am not convinced the problem is at deasync.

**Comment by denysvitali** [comment]
If request lost its context neither this should set myvalue to "bar" right?
```
function Foo(id)
{
    this.id = id;
    this.bar = function()
    {
        var done = false;
        var data;
        var myvalue = "foo";

        request("http://google.com/", function(err, res, body)
        {
            console.log("done1_in");
            //setTimeout(function(){done = true;}, 1); //workaround
            done = true;
            myvalue = "bar";
        });

        setTimeout(function()
        {
            console.log(myvalue);
            console.log("done1_out");
        },5000);
        return;
    };
}
```

Instead this actually works and the output is the expected:
```
done2_in
done2_out
done1_in
bar
done1_out
```
So I'm not really sure what problem is, but I'm thinking of a deasync bug.

**Comment by abbr** [comment]
Maybe losing context is not the correct explanation. Need more time to investigate. I also found if replacing `request` with node built-in [http.get](https://nodejs.org/api/http.html#http_http_get_options_callback) then there is no problem either. So the way `request` manipulating its callback is likely a contributing factor.

**Comment by denysvitali** [comment]
Maybe could be a .bind() issue? I'll check tomorrow...

**Comment by abbr** [comment]
Are you running on Windows?

**Comment by denysvitali** [comment]
Linux

**Comment by abbr** [comment]
What if you replace 
```
    while (!done)
    {
        deasync.runLoopOnce();
    }
```
with
```
    while (!done)
    {
        process._tickDomainCallback();
        if (!done) deasync.runLoopOnce();
    }
```

**Comment by denysvitali** [comment]
By replacing what you said it works like a charm.
Is that a deasync problem?

Anyway, I switched to Promise (bluebird) because best suits my needs

**Comment by abbr** [comment]
What's perplexing me is the updated code still hung in node@0.10. Had it not been so I would have thought I understand what goes wrong, concede it's a bug in `deasync`, and provide a fix. Now I am still unclear so I will refrain from drawing conclusion. `request` (or one of its components) definitely did something tricky on Node event loop interfering with `deasync`. Whatever fixes to-be implemented, regardless where the bug is, should also work in node@0.10.

**Comment by miliu99** [comment]
I can't make it work. It seems it's stuck in the while loop. Here is the code I tried for my mongoose query:

    companyModel.getQuoteOnOrBefore = function (symbol, time) {
        var histQuote;
        HistQuote.find({ symbol: symbol, time: { $lte: time } }).sort({ time: -1 }).limit(1).exec(function (err, results) {            
            setTimeOut(function () { histQuote = err ? null : (results[0] || null); }, 1);
        });
        while (histQuote === undefined) {
            deasync.runLoopOnce();
        }
        return histQuote;
    }

**Comment by renjithsasidharan** [comment]
First thanks for this awesome module. I have encontered a similar issue when using mongoose,
```

var item = mongoose.model('item');
    var done = false, output;
    var data = validator.data;
    item.find({store_id: data.store_id, _id: instance})
    .exec(function(err, result) {
      done = true;
      output = result;
    });
    while(done === false) {
      console.log("Waiting for async call..");
      require('deasync').runLoopOnce();
    }

```

**Comment by abbr** [comment]
@miliu99 @renjithsasidharan , Just published `deasync@0.1.0`. It solved the hung problem when   `request` module co-exists with `deasync` on Node 0.11+. But I haven't confirmed if mongoose has the same problem. You need to replace `runLoopOnce` in your code with `loopWhile`. See updated [readme](https://github.com/abbr/deasync#usages) for usage.
